components/net_traits/bluetooth_scanfilter.rs

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use device::bluetooth::BluetoothDevice;
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+const MAX_DEVICE_NAME_LENGTH: usize = 29;
+
+#[derive(Deserialize, Serialize)]
+pub struct ServiceUUIDSequence(Vec<String>);
+
+impl ServiceUUIDSequence {
+
+    pub fn new(vec: Vec<String>) -> ServiceUUIDSequence {
+        ServiceUUIDSequence(vec)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct BluetoothScanfilter {
+    name: String,
+    name_prefix: String,
+    services: ServiceUUIDSequence,
+}
+
+impl BluetoothScanfilter {
+    fn is_empty_or_invalid_filter(&self) -> bool {
+        self.name.is_empty() && self.name_prefix.is_empty() &&
+        self.services.is_empty() &&
+        self.name.len() > MAX_DEVICE_NAME_LENGTH &&
+        self.name_prefix.len() > MAX_DEVICE_NAME_LENGTH
+    }
+
+    pub fn new(name: String, name_prefix: String, services: Vec<String>) -> BluetoothScanfilter {
+        BluetoothScanfilter {
+            name: name,
+            name_prefix: name_prefix,
+            services: ServiceUUIDSequence::new(services),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct BluetoothScanfilterSequence(Vec<BluetoothScanfilter>);
+
+impl BluetoothScanfilterSequence {
+    pub fn new(vec: Vec<BluetoothScanfilter>) -> BluetoothScanfilterSequence {
+        BluetoothScanfilterSequence(vec)
+    }
+}
+
+impl BluetoothScanfilterSequence {
+    fn has_empty_or_invalid_filter(&self) -> bool {
+        self.0.is_empty() &&
+        self.0.iter().all(|x| !(x.is_empty_or_invalid_filter()))
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct RequestDeviceoptions {
+    // _filters: BluetoothScanfilterSequence,
```

Comment:
This code will be used after the current szeged/bluetoot-ipc get updated from servo/master.